### ISSUE TITLE
memoize map/fmap

### ISSUE THREAD

**Issue by acthp** [reactjs/reselect#62]
A common pattern in my app state is computed properties on dictionary elements. E.g. building an index of data that is identified by a key.

App state might look like

```
{ '1234-0001': arr1, '1234-0002': arr2, ... }
```

and the selector returns something like

```
{'1234-0001': makeExpensiveIndex(arr1), '1234-0002': makeExpensiveIndex(arr2), ... }
```

which might be done like ```_.mapObject(state, makeExpensiveIndex)```. You don't want to recompute the index on arr2 if only arr1 has changed.

At first glance it looks like reselect can't manage this case. Is there a simple way to do it? Or a different way of modeling dynamic data?

Similar patterns can happen with arrays, but it's harder to handle since we can't hash on object values in javascript. You have to either cache by index (position in the array), or by some unique property on the elements in the array (like the ```key``` prop in react element collections).

**Comment by mindjuice** [comment]
@acthp For now can you just use ``_.memoize()`` on your ``makeExpensiveIndex()`` function?

**Comment by acthp** [comment]
That will leak, of course. Might not be a problem, except for long sessions. I've written a ```fmapMemoize1``` that is like ```_.mapObject```, but caches the prop values from the previous call.

**Comment by mindjuice** [comment]
You could try: https://github.com/medikoo/memoizee

It lets you limit cache size on an LRU basis, and supports a WeakMap version to avoid hanging onto results for which the source objects no longer exist.

**Comment by kbrownlees** [comment]
I just hit a similar issue in my code base, I have solved it in the meantime by making a 'dynamic selector' which updates as the state changes. Comments on whether this is a good approach (/ safe) are welcome!

https://gist.github.com/kbrownlees/8ed41ed6d86499b8c590

I have a relatively stable state so I don't need to change it that often, but it makes the logic a lot nicer since you act on each 'stream' individually while still retaining the advantages of caching.

**Comment by arcanis** [comment]
I've started to experiment with the following code, what do you think? I'm somewhat mixed, because it dynamically adds a new property to some objects, which may hurt the JS engine optimizations, but API-wise, it seems quite elegant.

```js
let customComparator = (a, b) => {
    return a.length === b.length && a.every(aVal => b.find(bVal => aVal === bVal));
};

let getStreamModels = createSelector([
    (state, props) => useCustomCheck(customComparator, state.stream.map(locator => state.resourceRegistry.get(locator)),
], resources => {
    return resources;
})
```

```js
import { createSelectorCreator, defaultMemoize } from 'reselect';

let customCheckSymbol = Symbol();

export function useCustomCheck(check, value) {

    value[customCheckSymbol] = check;

    return value;

}

export function getCustomCheck(value) {

    return value[customCheckSymbol];

}

export function usesCustomCheck(value) {

    return Reflect.has(value, customCheckSymbol);

}

export let createSelector = createSelectorCreator(defaultMemoize, (a, b) => {

    if (a === b)
        return true;

    if (usesCustomCheck(a) && usesCustomCheck(b)) {

        let aCheck = getCustomCheck(a);
        let bCheck = getCustomCheck(b);

        if (aCheck !== bCheck)
            throw new Error();

        return aCheck(a, b);

    } else {

        return false;

    }

});
```

**Comment by HeyImAlex** [comment]
My intuition is that a _general_ solution that allows for nesting without _really_ changing reselect doesn't exist; reselect hides memoization from the end user by using pure functions, but a function that does sub-element caching is inherently impure.

However, writing a few common selector-creators for a single level (say map object property by key, array element by index, array element by key-function) wouldn't be too difficult, if a little inelegant. That would cover many use cases and wouldn't take a whole lot of effort.

Maybe an interesting line of thought; react does this (that is, nested memoization) very well. One of the things it _doesn't_ do well that we really need is control over evaluation. Maybe _combining_ the two systems could be a good solution? It's probably better suited as a new library, but maybe someone smarter than me wants to explore the idea a bit more.

**Comment by ellbee** [comment]
> My intuition is that a general solution that allows for nesting without really changing reselect doesn't exist; reselect hides memoization from the end user by using pure functions, but a function that does sub-element caching is inherently impure.

This is my intuition too.

> However, writing a few common selector-creators for a single level (say map object property by key, array element by index, array element by key-function) wouldn't be too difficult, if a little inelegant. That would cover many use cases and wouldn't take a whole lot of effort.

Agreed.

**Comment by ellbee** [comment]
> Unfortunately, I think you are correct, Reselect does not manage this case at the moment. I'm actually currently in the middle of a project where I ran into a similar issue. Soon, when I am less busy, I intend to look into the cases I am where I am aware that Reselect isn't a good fit, and either figure out a way to support them directly with Reselect, release some custom selectors with different memoization strategies or make some recipes for the docs.

Above is what I wrote (5 months ago!) in response to the OP. In the project I referenced the solution I ended up with was a custom selector with an LRU cache and a custom hashing scheme. At that point I had pretty much overridden all of Reselect with custom code.

In addition to the suggestions made by @HeyImAlex, maybe the docs could be improved with ideas of things to look at when the default Reselect stuff is not enough like using LRU caches, [making sure state shape is optimal](https://twitter.com/dan_abramov/status/720219615041859584) etc.

I will get around to this eventually, but in the mean time if anyone has any ideas for common selector-creators or wants to take a shot at the docs it would be welcome.

**Comment by arcanis** [comment]
The approach detailed above (separating the data selection from the data comparison so that the compared data may be different from the data actually returned by the selector) seems to work quite well for our use case. Do you see an issue with it?

**Comment by ellbee** [comment]
Sorry @arcanis, I somehow missed it. I will have a play with it.

**Comment by HeyImAlex** [comment]
@arcanis I'm not sure that approach addresses the issue? It's a little hard to follow from my phone, but it seems like that allows you to use a per-dependency equalityCheck function. If you applied it to the code in the original issue, we'd still be recomputing for every element. Unless I've misunderstood?

**Comment by HeyImAlex** [comment]
I just finished up a separate package, [reselect-map](https://github.com/HeyImAlex/reselect-map), with the new selector creators we had talked about. Because it's so niche and the code ended up being more complex than expected, I wasn't sure it was a great candidate for being folded into reselect, but I'm not opposed or anything.

RE: Documentation, I feel like this issue should just be left to languish and those who *really need it* will eventually land here. Adding an **advanced reselect** section to the readme will mean a bunch of people implementing optimizations they don't need. But then again, maybe it's not cool for us to make that call...

**Comment by ellbee** [comment]
Nice work! 

In my opinion it makes sense to keep `reselect-map` as a separate package. I think the majority of Reselect users only need to use `createSelector`, so it is good to keep the API small.

> RE: Documentation, I feel like this issue should just be left to languish and those who really need it will eventually land here.

Yes, you are probably right, and the documentation for Reselect is already really long for such a tiny library.

> Adding an advanced reselect section to the readme will mean a bunch of people implementing optimizations they don't need.

Agreed. This happened with the new documentation I wrote about the `mapStateToProps` optimisation in React Redux.

**Comment by mindjuice** [comment]
Agree that a separate package makes more sense.

IMO though, it might be worth a small section in `reselect`'s `README.md` to call out the use case for `reselect-map`, but with an admonition that you very likely don't need it (i.e., if you can't understand the use cases, you don't need it).  Or even just a "Related Projects" section with a link and little/no description.  Just have the (existing) admonitions in the `reselect-map` repo.

**Comment by ellbee** [comment]
Yes, fair point.

**Comment by ellbee** [comment]
Added a section about reselect-map to the README. Thanks for the input everyone.
